onnx_trt_backend.cpp

```diff
@@ -0,0 +1,771 @@
+    break;
+    case ONNXIFI_BACKEND_EXTENSIONS:
+    *infoValueSize = 0;
+    break;
+    case ONNXIFI_BACKEND_DEVICE:
+    SET_STRING("gpu");
+    break;
+    case ONNXIFI_BACKEND_DEVICE_TYPE:
+    SET_UINT64(ONNXIFI_DEVICE_TYPE_GPU);
+    break;
+    case ONNXIFI_BACKEND_CAPABILITIES:
+    SET_UINT64(0UL);
+    break;
+    case ONNXIFI_BACKEND_INIT_PROPERTIES:
+     SET_UINT64(0UL);
+     break;
+    case ONNXIFI_BACKEND_MEMORY_TYPES:
+     SET_UINT64(ONNXIFI_MEMORY_TYPE_CUDA_BUFFER);
+     break;
+    case ONNXIFI_BACKEND_MEMORY_SIZE:
+     {
+      size_t free, total;
+      if (cudaMemGetInfo(&free, &total) != cudaSuccess) {
+        return ONNXIFI_STATUS_BACKEND_UNAVAILABLE;
+      }
+      SET_UINT64(uint64_t(total));
+      break;
+     }
+    // TODO: Dummy numbers below
+    case ONNXIFI_BACKEND_MAX_GRAPH_SIZE:
+     SET_UINT64(1000000UL);
+     break;
+    case ONNXIFI_BACKEND_MAX_GRAPH_COUNT:
+     SET_UINT64(1UL);
+     break;
+    case ONNXIFI_BACKEND_MACS_FP32:
+     SET_UINT64(0UL);
+     break;
+    case ONNXIFI_BACKEND_MACS_FP16:
+     SET_UINT64(0UL);
+     break;
+    case ONNXIFI_BACKEND_MEMORY_BANDWIDTH:
+     SET_UINT64(0UL);
+     break;
+    case ONNXIFI_BACKEND_CPU_MEMORY_READ_BANDWIDTH:
+     SET_UINT64(0UL);
+     break;
+    case ONNXIFI_BACKEND_CPU_MEMORY_WRITE_BANDWIDTH:
+     SET_UINT64(0UL);
+     break;
+    default:
+     return ONNXIFI_STATUS_UNSUPPORTED_PARAMETER;
+    }
+    return ONNXIFI_STATUS_SUCCESS;
+#undef SET_STRING
+#undef SET_UINT64
+  });
+}
+
+ONNXIFI_PUBLIC ONNXIFI_CHECK_RESULT onnxStatus ONNXIFI_ABI ONNXIFI_SYMBOL_NAME(
+    onnxGetBackendCompatibility)(onnxBackendID backendID, size_t onnxModelSize,
+                                 const void *onnxModel) {
+  return OnnxifiTryCatch([&] {
+    if (!onnxModel) {
+      return ONNXIFI_STATUS_INVALID_POINTER;
+    }
+    if (onnxModelSize == 0) {
+      return ONNXIFI_STATUS_INVALID_SIZE;
+    }
+
+    // NB: not ideal case. We CHECK model by actually trying to run the
+    // conversion. However, this might be the case for other vendors
+    auto *backend_id = reinterpret_cast<OnnxTensorRTBackendID *>(backendID);
+    if (!backend_id) {
+      return ONNXIFI_STATUS_INVALID_ID;
+    }
+    OnnxTensorRTBackendRep backendrep(*backend_id);
+    return backendrep.ImportModel(onnxModel, onnxModelSize, 0, nullptr);
+  });
+}
+
+// NB: Passing arguments to backend is tricky. And we need more documentation
+// for it I didn't put any arguments here for now.
+// TODO: submit arguments for
+// - setMaxBatchSize (size_t)
+// - setMaxWorkspaceSize (size_t)
+// - setHalf2Mode (bool)
+// - setInt8Mode (bool)
+// - setDebugSync (bool)
+ONNXIFI_PUBLIC ONNXIFI_CHECK_RESULT onnxStatus ONNXIFI_ABI ONNXIFI_SYMBOL_NAME(
+    onnxInitBackend)(onnxBackendID backendID, const uint64_t *auxPropertiesList,
+                     onnxBackend *backend) {
+  return OnnxifiTryCatch([&] {
+    auto *backend_id = reinterpret_cast<OnnxTensorRTBackendID *>(backendID);
```

Comment:
`onnxInitBackend`, `onnxInitGraph`, `onnxInitEvent` must set backend/graph/event to NULL if they fail.